Update docs to use weaver for code generation (#1296)
open-telemetry/semantic-conventions: docs/non-normative/code-generation.md

- Deprecated conventions that reached stability should not be removed without major version update according to SemVer.
- Conventions that were deprecated while being experimental should still be generated and kept in the preview (part of) semantic conventions artifact. It minimizes runtime issues
  and breaking changes in user applications.

Keep stable convention definitions inside the preview (part of) semantic conversions artifact. It prevents user code from breaking when semantic convention stabilizes. Deprecate stable definitions inside the preview artifact and point users to the stable location in generated documentation.
For example, in Java `http.request.method` attribute is defined as the deprecated `io.opentelemetry.semconv.incubating.HttpIncubatingAttributes.HTTP_REQUEST_METHOD` field and also as stable `io.opentelemetry.semconv.HttpAttributes.HTTP_REQUEST_METHOD`.

## Semantic Conventions Artifact Structure

This section contains suggestions on structuring semantic convention artifact(s) which should be adjusted to the specific language.

- Artifact name:
  - `opentelemetry-semconv` - stable conventions

>   - `opentelemetry-semconv-incubating` - (if applicable) the preview artifact containing all conventions

Author: yep, clarified

- Namespace: `opentelemetry.semconv` and `opentelemetry.semconv.incubating`
- All supported Schema URLs should be listed to allow different instrumentations in the same application to provide the exact version of conventions they follow.
- Attributes, metrics, and other convention definitions should be grouped by the convention type and the root namespace. See the example below:

```

> ├── SchemaUrls.code

Reviewer: Soon we will hopefully be able to use mermaid diagrams also for folder structure https://github.com/mermaid-js/mermaid/pull/5665 🙏

├── attributes
│   ├── ClientAttributes.code
│   ├── HttpAttributes.code
│   └── ...
├── metrics
│   ├── HttpMetrics.code
│   └── ...
└── events
    └── ...
```

## Generating semantic conventions

This section describes how to do code-generation with weaver.

> > Note:

Reviewer: ```suggestion
> [!IMPORTANT] 
```

That generates a nice callout:

![image](https://github.com/user-attachments/assets/9cfd9568-4a46-4083-9210-eff988248b58)

> We're transitioning from [build-tools](https://github.com/open-telemetry/build-tools/blob/main/semantic-conventions/README.md#code-generator)
> to [opentelemetry-weaver](https://github.com/open-telemetry/weaver/blob/main/crates/weaver_forge/README.md) to generate code for semantic conventions.
> All new code-generation should be done using weaver, build-tools may become incompatible with future version of semantic conventions.

> Code-generation is based on YAML definitions in the specific version of semantic conventions and usually

Reviewer: ```suggestion
Code-generation is based on YAML definitions in the specific version of semantic conventions.
Usually, it involves several steps where some can be semi-automated:
```

To avoid the sentence to be too long :)

involves several steps which could be semi-automated:

1. Manually update the Semantic Conventions version in config
2. Add the new Schema URL to the list of supported versions
   - If it's not automated, then it can, at least, be automatically checked.
3. Check out (or download) the new version of Semantic Conventions
4. Run code-generation script (see below for the details)
5. Fix lint violations in the auto-generated code (if any)
6. Send the PR with new code to the corresponding repository

Here are examples of how steps 2-5 are implemented for [Python](https://github.com/open-telemetry/opentelemetry-python/pull/4091) and [Erlang](https://github.com/open-telemetry/opentelemetry-erlang/pull/733).

Step 4 (running code generation) depends on language-specific customizations. It's also the only step that's affected by tooling migration.

Check out[weaver code-generation documentation for the details](https://github.com/open-telemetry/weaver/blob/main/crates/weaver_forge/README.md)

### Migrating from build-tools

> Migration from build-tools involves changing Jinja templates and adding [weaver config file](https://github.com/open-telemetry/weaver/blob/main/crates/weaver_forge/README.md#configuration-file---weaveryaml).

Reviewer: ```suggestion
Migration from build-tools involves changing Jinja templates and adding a [weaver config file](https://github.com/open-telemetry/weaver/blob/main/crates/weaver_forge/README.md#configuration-file---weaveryaml).
```

#### Weaver config

Here's a simplified example of this file that generates all attributes.

```yaml
params:
  excluded_namespaces: [ios, aspnetcore, signalr, android, dotnet, jvm, kestrel]

templates:
  - pattern: semantic_attributes.j2
    filter: >
      semconv_grouped_attributes({
        "exclude_root_namespace": $excluded_namespaces
      })
      | map({
          root_namespace: .root_namespace,
          attributes: .attributes,
          output: $output

>         })

Reviewer: Do we really need this with the new filters Laurent provided? In my .NET current WIP Code, I get all of this already with this simple weaver config file:

```
templates:
  - pattern: SemanticConventionsAttributes.cs.j2
    filter: >
      semconv_grouped_attributes
    application_mode: each
```

Reviewer 2: I agree. I'm not sure that most of the codegen authors will need the output field, which is specific to the Python codegen.

Author: my goal was to demonstrate that you can  post-process and add more things into the context.

The reason python does it is because it generates metrics in addition to attributes and needs an extra folder structure for it. I think everyone would need some form of it.

I was a bad influence and now erlang does it too https://github.com/open-telemetry/opentelemetry-erlang/pull/733/files#diff-1a4c4f5071244fadabb10f3aed3958e7e4486869f3260201dc70509330fdc082

Reviewer 2: In that case, why not directly use the `$output` parameter from the template (i.e. ctx.params.output)? I don't see the value in including it from a JQ projection.

Author: I updated the example to be more realistic and to modify the output `output: $output + "attributes/"`

    application_mode: each
```

You can configure language-specific parameters in the `params` section of the config or pass them with `--param key=value` arguments when

> running weaver command from the code generation script (similarly to build-tools `-Dparam=value` option).

Reviewer 2: `-Dparam=value` is also supported in Weaver.

Weaver is able to run code-generation for multiple templates (defined in the corresponding section) at once.

Before executing Jinja, weaver allows to filter or process semantic convention definitions in `filter` section for each template.

> In this example it uses `semconv_grouped_attributes` filter - a helper method that groups attribute definitions by root namespace and excludes

Reviewer: ```suggestion
In this example, it uses `semconv_grouped_attributes` filter - a helper method that groups attribute definitions by root namespace and excludes
```

attributes not relevant to this language. You can write alternative or additional filters and massage semantic conventions data using [JQ](https://jqlang.github.io/jq/manual/).

> The `application_mode: each` configures weaver to run code generation for each semconv group and, as a consequence, generate code for each group
> in a different file.

Reviewer 2: ```suggestion
The `application_mode: each` configures weaver to run code generation for each semconv group and, as a consequence, generate code for each group
in a different file. The application mode `single` is also supported to apply the template to all groups at once.
```

See

> [weaver code-generation docs](https://github.com/open-telemetry/weaver/blob/main/crates/weaver_forge/README.md) for the details on the config, data schema, JQ filters, and more.

Reviewer: ```suggestion
[weaver code-generation docs](https://github.com/open-telemetry/weaver/blob/main/crates/weaver_forge/README.md) 
for the details on the config, data schema, JQ filters, and more.
```

#### Jinja templates

> Jinja templates need to be changed to leverage (better) data structure and helper methods.
> The first key difference is that each jinja template should define how to name the corresponding file(s).

Reviewer 2: ```suggestion
Jinja templates need to be changed to leverage (better) data structure and helper methods.
The first key difference is that each jinja template can define how to name the corresponding file(s). If you
don't specify the name of the output file via the method `set_file_name`, Weaver will use the relative path
and the name of the template itself to determine the output file.
```

Experimental semantic conventions are intended for end-user applications.

### Deprecated Conventions

It's recommended to generate code for deprecated attributes, metrics, and other conventions. Use appropriate annotations to mark them as deprecated.
Conventions have a `stability` property which provide the stability level at the deprecation time (`experimental` or `stable`) and
the `deprecated` property that describes deprecation reason which can be used to generate documentation.

- Deprecated conventions that reached stability should not be removed without major version update according to SemVer.
- Conventions that were deprecated while being experimental should still be generated and kept in the preview (part of) semantic conventions artifact. It minimizes runtime issues
  and breaking changes in user applications.

Keep stable convention definitions inside the preview (part of) semantic conversions artifact. It prevents user code from breaking when semantic convention stabilizes. Deprecate stable definitions inside the preview artifact and point users to the stable location in generated documentation.

> For example, in Java, the attribute `http.request.method` is defined as the deprecated in both stable and preview artifacts (e.g., `io.opentelemetry.semconv.incubating.HttpIncubatingAttributes.HTTP_REQUEST_METHOD`, `io.opentelemetry.semconv.HttpAttributes.HTTP_REQUEST_METHOD`).

Reviewer: ```suggestion
For example, in Java, the attribute `http.request.method` is defined as deprecated in both stable and preview artifacts (e.g., `io.opentelemetry.semconv.incubating.HttpIncubatingAttributes.HTTP_REQUEST_METHOD`, `io.opentelemetry.semconv.HttpAttributes.HTTP_REQUEST_METHOD`).
```

I messed up my previous suggestion :D
